File: src/main/java/io/jjong/algorithm/word/ReverseWord.java
package io.jjong.algorithm.word;

/**
 * create on 2022/12/03. create by IntelliJ IDEA.
 *
 * <p> 클래스 설명 </p>
 * <p> {@link } and {@link }관련 클래스 </p>
 *
 * @author Jongsang Han(Henry)
 * @version 1.0
 * @see
 * @since 1.0
 */
public class ReverseWord {
  public static void main(String[] args){
    String word = "jonsang is king is power";
    char[] charArray = word.toCharArray();
    reverseWord(charArray);

    System.out.println(charArray);
  }

  public static void reverseWord(char[] input) {
    int wordLength = input.length;

    // 전체를 뒤집는다.
    reverse(input, 0, wordLength-1);
    System.out.print("reverse : ");
    System.out.println(input);

    // 문자열의 각 단어를 뒤집는다.
    int start = 0, end = 0;
    while (start < wordLength) {
      // 문자열 읽을 index 구하기
      // 시작은 공백일 경우
      while (start < end || (start < wordLength && input[start] == ' ')) {
        System.out.print("start : " + (start+1));
        System.out.println(", char : " + input[start]);
        start++;
      }
      // 끝은 공백이 아닐 경우
      while (end < start || (end < wordLength && input[end] != ' ')) {
        System.out.print("end : " + (end + 1));
        System.out.println(", char : " + input[end]);
        end++;
      }
      reverse(input, start, end-1);
    }
  }

  private static void reverse(char[] input, int start, int end) {
    while (start < end) {
      char tmp = input[start];
      input[start++] = input[end];
      input[end--] = tmp;
    }
  }

  private static int find(char[] input, char c, int start) {
    for (int i=start; i<input.length; i++) {
      if (input[i] == c) {
        return i;
      }
    }
    return -1;
  }

}

File: archive/src/main/kotlin/io/jjong/algorithm/word/ReverseWord.kt
package io.jjong.algorithm.word

/**
 * 단어 단위로 순서를 뒤집는다. 전체를 한 번 뒤집은 뒤 각 단어를 다시 뒤집는 고전적 기법.
 * 예: "abc de" → "de abc". 입력 [CharArray]를 제자리(in-place)에서 변경한다.
 */
object ReverseWord {
    fun reverseWord(input: CharArray) {
        val wordLength = input.size

        // 전체를 뒤집는다.
        reverse(input, 0, wordLength - 1)
        print("reverse : ")
        println(String(input))

        // 각 단어를 다시 뒤집는다.
        var start = 0
        var end = 0
        while (start < wordLength) {
            // 시작 위치: 공백을 건너뛴다.
            while (start < end || (start < wordLength && input[start] == ' ')) {
                print("start : ${start + 1}")
                println(", char : ${input[start]}")
                start++
            }
            // 끝 위치: 공백이 아닌 동안 전진한다.
            while (end < start || (end < wordLength && input[end] != ' ')) {
                print("end : ${end + 1}")
                println(", char : ${input[end]}")
                end++
            }
            reverse(input, start, end - 1)
        }
    }

    private fun reverse(input: CharArray, start: Int, end: Int) {
        var s = start
        var e = end
        while (s < e) {
            val tmp = input[s]
            input[s++] = input[e]
            input[e--] = tmp
        }
    }
}

fun main() {
    val word = "jonsang is king is power"
    val charArray = word.toCharArray()
    ReverseWord.reverseWord(charArray)

    println(String(charArray))
}
